File: src/data/raytrace/PropertyOnLineCalculator.java
package data.raytrace;

import java.awt.Color;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import data.DataHandler;
import data.raytrace.RaytraceScene.RaySimulationObject;
import data.raytrace.raygen.RayGenerator;
import geometry.Geometry;
import geometry.Geometry.NearestPointCalculator;
import geometry.Vector3d;
import io.Drawer;
import jcomponents.raytrace.RaySimulationData;
import maths.Controller;
import maths.Operation.CalculationController;
import maths.data.ArrayOperation;
import util.ArrayUtil;
import util.JFrameUtils;
import util.data.DoubleArrayList;

public class PropertyOnLineCalculator {
	private static final Logger logger = LoggerFactory.getLogger(PropertyOnLineCalculator.class);
	private RaytraceScene scene;
	public final DoubleArrayList dataPoints = new DoubleArrayList();

	public static enum VisualizationMode
	{
	    REFRACTIVE_INDEX, FOCAL_DISTANCE, DIVERGENCE;
	}


	public PropertyOnLineCalculator(RaytraceScene scene)
	{
		this.scene = scene;
	}

	public void paint(Vector3d position, Vector3d direction, double rangeBegin, double rangeEnd, VisualizationMode mode, boolean sphereArc, Drawer drawer)
	{
		dataPoints.clear();
		drawer.setColor(Color.BLACK);
		//ArrayList<IntersectionPoint> list = new ArrayList<>();
		ArrayList<OpticalObject> objects = new ArrayList<>();
		DoubleArrayList rayObjectIntersectionDistances = new DoubleArrayList();
		DoubleArrayList focusIntersectionDistances = new DoubleArrayList();
		OpticalObject nextSurfaces[] = scene.copyActiveSurfaces();
		OpticalObject nextVolumes[] = scene.getActiveVolumes();
		OpticalObject nextMeshes[] = scene.getActiveMeshes();

        RaytraceScene.RaySimulationObject ray = new RaytraceScene.RaySimulationObject();
        Intersection intersection = ray.nearest;
		intersection.distance = 1e-10;
		double width = drawer.getWidth(), height = drawer.getHeight();
		direction.normalize();
		//Get 16 additional positions in a circle around position on the normal plane of direction
		Vector3d[] positions = new Vector3d[16];
		Vector3d[] orthonormals = new Vector3d[2];
        for (int i = 0; i < 2; ++i)
        {
            orthonormals[i] = new Vector3d();
        }

		Geometry.getOrthorgonalVectors(direction, orthonormals[0], orthonormals[1]);
		for (int i = 0; i < 16; ++i)
        {
            double angle = i * Math.PI / 8;
            positions[i] = new Vector3d(position);
            positions[i].add(orthonormals[0], Math.cos(angle));
            positions[i].add(orthonormals[1], Math.sin(angle));
        }
		switch (mode)
		{
		    case REFRACTIVE_INDEX:
            {
			    NearestPointCalculator npc = new NearestPointCalculator(3);
                Vector3d refracted_direction = new Vector3d();
                Vector3d focalPoint = new Vector3d();
				while(true)
				{
					double lastDist = intersection.distance;
					for (int i = 0; i < 16; ++i)
                    {
                        intersection.distance = Double.POSITIVE_INFINITY;
                        intersection.object = null;
                        RaytraceScene.getNextIntersection(
                                positions[i],
                                direction,
                                intersection,
                                lastDist + 1e-10,
                                nextSurfaces,
                                nextVolumes,
                                nextMeshes);
                        if (intersection.object instanceof SurfaceObject)
                        {
                            refracted_direction.set(direction);
                            scene.apply_surface_to_ray(ray, null, -1, null, -1, intersection, refracted_direction);
                            npc.addRay(intersection.position, refracted_direction, 0);
                        }
                    }
                    npc.calculate();
                    npc.get(focalPoint);
                    focusIntersectionDistances.add(direction.dot(focalPoint));

					intersection.distance = Double.POSITIVE_INFINITY;
					intersection.object = null;

					RaytraceScene.getNextIntersection(
								position,
								direction,
								intersection,
								lastDist + 1e-10,
								nextSurfaces,
								nextVolumes,
								nextMeshes);
					rayObjectIntersectionDistances.add(intersection.distance);
					OpticalObject object = intersection.object;
					if (object == null || objects.size() > 100)
					{
					     break;
					}
					objects.add(object);

					nextSurfaces = object.surfaceSuccessor;
					nextVolumes = object.volumeSuccessor;
					nextMeshes = object.meshSuccessor;
					npc.reset();
				}
				CalculationController control = new Controller();
				System.out.println(Arrays.toString(objects.toArray(new OpticalObject[objects.size()])));
				System.out.println(Arrays.toString(focusIntersectionDistances.toArrayD()));
				for (int i = 0; i < objects.size(); ++i)
				{
					OpticalObject obj = objects.get(i);
					if (obj instanceof OpticalSurfaceObject)
					{
						OpticalSurfaceObject oso = (OpticalSurfaceObject)obj;
						double ior0, ior1;
						if (oso.direction.dot(direction) < 0)
						{
							ior0 = oso.ior0.calculate(scene.vs, control).doubleValue();
							ior1 = oso.ior1.calculate(scene.vs, control).doubleValue();
						}
						else
						{
							ior0 = oso.ior1.calculate(scene.vs, control).doubleValue();
							ior1 = oso.ior0.calculate(scene.vs, control).doubleValue();
						}
						double xLoc = rayObjectIntersectionDistances.getD(i);
						dataPoints.add(xLoc);
						dataPoints.add(ior0);
						dataPoints.add(xLoc);
						dataPoints.add(ior1);
						ior0 = (2 - ior0) * drawer.getHeight() / 2;
						ior1 = (2 - ior1) * drawer.getHeight() / 2;
						xLoc = (xLoc - rangeBegin) / (rangeEnd - rangeBegin) * width;
						drawer.pushPoint(xLoc, ior0);
						drawer.pushPoint(xLoc, ior1);
					}
					else if (obj instanceof OpticalVolumeObject)
					{
						OpticalVolumeObject ovo = (OpticalVolumeObject)obj;
						double start = rayObjectIntersectionDistances.getD(i);
						double steplength = (rayObjectIntersectionDistances.getD(i + 1) - start)/99.;
						for (int j = 0; j < 100; ++j)
						{
							double mult = start + j * steplength;
							double ior = ovo.getRefractiveIndex(position.x + direction.x * mult, position.y + direction.y * mult, position.z + direction.z * mult);
							dataPoints.add(mult);
							dataPoints.add(ior);
							double xLoc = mult;
							xLoc = (xLoc - rangeBegin) / (rangeEnd - rangeBegin) * width;
							drawer.pushPoint(xLoc, (2 - ior) * drawer.getHeight() / 2);
						}
					}
				}
				try {
					drawer.drawPolyLine();
				} catch (IOException ex) {
					JFrameUtils.logErrorAndShow("Can't update Visualisation", ex, logger);
				}
			break;
			}
            case FOCAL_DISTANCE:
            {
                int num_rays = 3000;
                int num_evaluations = 200;
                ArrayList<OpticalObject> oso = new ArrayList<>();
                scene.getActiveLights(oso);
                RayGenerator gen = new RayGenerator();
                RaySimulationObject rso = new RaySimulationObject();
                RaySimulationData rsd = new RaySimulationData(num_rays, false);
                Vector3d npcDir = new Vector3d();
                Vector3d npcPos = new Vector3d();
                Vector3d npcCenter = new Vector3d();
                int maxBounces = 30;
                for (int i = 0; i < oso.size(); ++i)
                {
                    gen.threeDimensional = true;
                    gen.setSource(oso.get(i));
                    double[] trajectory = new double[num_rays * (maxBounces + 2) * 3];
                    Arrays.fill(trajectory, Double.NaN);
                    scene.calculateRays(0, num_rays, num_rays, gen, 0, 0, null, null, rsd.endpoints, rsd.enddirs, rsd.endcolor, trajectory , rsd.accepted, rsd.bounces, rsd.lastObject, maxBounces, false, rso, RaytraceScene.UNACCEPTED_MARK);
                    double distances[] = ArrayUtil.fillEquidistant(rangeBegin, rangeEnd, new double[num_evaluations]);
                    final NearestPointCalculator npc = new NearestPointCalculator(3);
                    double result[] = new double[distances.length];
                    double trajectoryDistances[][] = new double[num_rays][];

                    for (int iRay = 0; iRay < num_rays; ++iRay)
                    {
                        double tmp[] = new double[maxBounces * 3 + 6];
                        for (int iBounce = 0; iBounce < maxBounces; ++iBounce)
                        {
                            tmp[iBounce] = direction.dot(trajectory, (iRay * (maxBounces + 2) + iBounce) * 3);
                            if (Double.isNaN(tmp[iBounce]) || iBounce + 1== maxBounces)
                            {
                                trajectoryDistances[iRay] = Arrays.copyOf(tmp, iBounce);
                                break;
                            }
                        }
                    }
                    dataPoints.clear();

                    for (int iDist = 0; iDist < distances.length; ++iDist)
                    {
                        for (int iRay = 0; iRay < num_rays; ++iRay)
                        {
                            int idx = Arrays.binarySearch(trajectoryDistances[iRay], distances[iDist]);
                            if (idx < 0){
                                idx = -idx - 1;
                            }
                            if (idx >= trajectoryDistances[iRay].length || idx < 1)
                            {
                                continue;
                            }
                            if (distances[iDist] < trajectoryDistances[iRay][idx - 1] || distances[iDist] > trajectoryDistances[iRay][idx])
                            {
                                throw new RuntimeException();
                            }
                            npcDir.set(trajectory, (iRay * (maxBounces + 2) + idx) * 3);
                            npcPos.set(trajectory, (iRay * (maxBounces + 2) + idx - 1) * 3);
                            npcDir.sub(npcPos);
                            npcDir.normalize();
                            //System.out.println(npcDir + " " + npcPos);
                            npc.addRay(npcPos, npcDir, 0);
                        }

                        dataPoints.add(distances[iDist]);
                        if (npc.calculate() == 3 && npc.getUniformness() < 0.999)
                        {
                            npc.get(npcCenter);
                            //System.out.println(npcCenter);
                            result[iDist] = direction.dot(npcCenter);
                        }
                        else
                        {
                            result[iDist] = Double.NaN;
                        }

                        dataPoints.add(result[iDist]);
                        npc.reset();
                    }
                    double max = ArrayUtil.max(result);
                    double multY = height / max;
                    double multX = width / result.length;
                    for (int iDist = 0; iDist < distances.length; ++iDist)
                    {
                        drawer.pushPoint(iDist * multX, height - multY * result[iDist]);
                    }
                    try {
                        drawer.drawPolyLine();
                    } catch (IOException ex) {
                        JFrameUtils.logErrorAndShow("Can't update Visualisation", ex, logger);
                    }
                    DataHandler.globalVariables.setGlobal("focalpoints", new ArrayOperation(result));
                    System.out.println("result" + Arrays.toString(result));
                }
                break;
            }
			case DIVERGENCE:
			{
				int num_rays = 1000;
				int num_evaluations = 200;
				ArrayList<OpticalObject> oso = new ArrayList<>();
				scene.getActiveLights(oso);
				RayGenerator gen = new RayGenerator();
				RaySimulationObject rso = new RaySimulationObject();
				RaySimulationData rsd = new RaySimulationData(num_rays, false);
				for (int i = 0; i < oso.size(); ++i)
				{
					gen.threeDimensional = true;
					gen.setSource(oso.get(i));
					scene.calculateRays(0, num_rays, num_rays, gen, 0, 0, null, null, rsd.endpoints, rsd.enddirs, rsd.endcolor, null, rsd.accepted, rsd.bounces, rsd.lastObject, 10, false, rso, RaytraceScene.UNACCEPTED_MARK);

					double distances[] = ArrayUtil.fillEquidistant(rangeBegin, rangeEnd, new double[num_evaluations]);
					double result[] = sphereArc ? Geometry.getVarianceOnSphere(rsd.endpoints, rsd.enddirs, rsd.accepted, position, direction, distances, new double[num_evaluations])
							 : Geometry.getVariance(rsd.endpoints, rsd.enddirs, rsd.accepted, position, direction, distances, new double[num_evaluations]);
					ArrayUtil.sqrt(result, 0, result.length);
					double max = ArrayUtil.max(result);
					double multY = height / max;
					double multX = width / result.length;
					dataPoints.clear();
					for (int j = 0; j < result.length; ++j)
					{
						dataPoints.add(distances[j]);
						dataPoints.add(result[j]);
						drawer.pushPoint(j * multX, height - multY * result[j]);
					}
					try {
						drawer.drawPolyLine();
					} catch (IOException ex) {
						JFrameUtils.logErrorAndShow("Can't update Visualisation", ex, logger);
					}
					DataHandler.globalVariables.setGlobal("divergence", new ArrayOperation(result));
					System.out.println("result" + Arrays.toString(result));
				}
				break;
			}
		}
	}
}

File: src/data/raytrace/PropertyOnLineCalculator.py
import numpy as np
from enum import Enum
import logging

from calgraph3d.data.raytrace.RaytraceScene import RaytraceScene
from calgraph3d.data.raytrace.Intersection import Intersection
from calgraph3d.data.raytrace.OpticalSurfaceObject import OpticalSurfaceObject
from calgraph3d.data.raytrace.OpticalVolumeObject import OpticalVolumeObject
from calgraph3d.data.raytrace.MeshObject import MeshObject
from jsymmath.geometry.Geometry import NearestPointCalculator
from jsymmath.util import ArrayUtil


logger = logging.getLogger(__name__)


class VisualizationMode(Enum):
    REFRACTIVE_INDEX = 1
    FOCAL_DISTANCE = 2
    DIVERGENCE = 3

class PropertyOnLineCalculator:
    def __init__(self, scene: RaytraceScene):
        self.scene = scene
        self.lineEvaluationDistances = None
        self.lineRefractiveIndices = None
        self.lineObjects = None
        self.bundleIntersectionDistances = None
        self.bundleRearPrincipalPoints = None
        self.bundleEvaluationDistances = None
        self.xp = None

    def switchToLibrary(self, xp):
        self.xp = xp
        self.lineEvaluationDistances = ArrayUtil.convert(self.lineEvaluationDistances, xp)
        self.lineRefractiveIndices = ArrayUtil.convert(self.lineRefractiveIndices, xp)
        self.bundleIntersectionDistances = ArrayUtil.convert(self.bundleIntersectionDistances, xp)
        self.bundleRearPrincipalPoints = ArrayUtil.convert(self.bundleRearPrincipalPoints, xp)
        self.bundleEvaluationDistances = ArrayUtil.convert(self.bundleEvaluationDistances, xp)
        self.lineObjects = ArrayUtil.convert(self.lineObjects, xp)

    def interpolate_rear_principal_point(self, x):
        return self.xp.interp(x, self.bundleEvaluationDistances, self.bundleRearPrincipalPoints)

    def interpolate_bundle_intersection_point(self, x):
        return self.xp.interp(x, self.bundleEvaluationDistances, self.bundleIntersectionDistances)

    def interpolate_ior(self, x):
        return self.xp.interp(x, self.lineEvaluationDistances, self.lineRefractiveIndices)

    def get_dioptres(self, measurement_point):
        dioptre = 1000 / (self.bundleIntersectionDistances - measurement_point)
        self.xp.nan_to_num(dioptre, nan=0, copy=False)
        return dioptre

    def getLineInformation(
            self,
            start_position: np.ndarray,
            start_direction: np.ndarray,
            rangeBegin: float,
            rangeEnd: float,
            mode: VisualizationMode,
            num_eval: int=1000,
            spread: float=0.1,
            sphereArc: bool=True):
        xp = ArrayUtil.getArrayModule(start_position)
        self.xp = xp
        start_direction = start_direction / xp.linalg.norm(start_direction)
        nextSurfaces = self.scene.optical_surface_objects
        nextVolumes = self.scene.optical_volume_objects
        meshSurfaces = self.scene.optical_mesh_objects



        # ---- REFRACTIVE_INDEX mode ----
        if mode == VisualizationMode.REFRACTIVE_INDEX:
            start_position = start_position.reshape(1, 3)
            start_direction = start_direction.reshape(1, 3)
            object_ids = []
            rayDistances = []
            refractiveIndices = []
            locations = []
            resultObjects = []
            lowerBound = xp.full(shape=(1,), fill_value=rangeBegin, dtype=start_position.dtype)

            while True:
                intersection = Intersection(shape=1, xp=xp)
                intersection.distance[:] = rangeEnd
                RaytraceScene.getNextIntersection(
                    start_position,
                    start_direction,
                    intersection,
                    lowerBound,
                    nextSurfaces,
                    nextVolumes,
                    meshSurfaces
                )
                assert (intersection.distance + 0.01 >= lowerBound).all(), f"Intersection distance {intersection.distance} is less than lower bound {lowerBound}"
                lowerBound[:] = intersection.distance + 0.01
                obj = intersection.object
                if obj[0] == -1 or len(object_ids) > 100:
                    break
                rayDistances.append(intersection.distance.copy())
                object_ids.append(obj)
                opt_object = self.scene.getObjectById(int(obj[0]))
                if isinstance(opt_object, OpticalVolumeObject):
                    opt_object.getIntersection(
                        start_position,
                        start_direction,
                        intersection,
                        lowerBound,
                        xp.full(shape=(1,), fill_value=rangeEnd, dtype=start_position.dtype),
                        xp=xp, enterVolume=False)
                    lowerBound[:] = intersection.distance
                    rayDistances.append(intersection.distance)
                    object_ids.append(obj)


                successors = self.scene.get_successors(self.scene.getObjectById(int(obj[0])))
                nextSurfaces = [s for s in successors if isinstance(s, OpticalSurfaceObject)]
                nextVolumes = [s for s in successors if isinstance(s, OpticalVolumeObject)]
                meshSurfaces = [s for s in successors if isinstance(s, MeshObject)]
            rayDistances = xp.concatenate(rayDistances)
            object_ids = xp.concatenate(object_ids)
            print("rayDistances", " ".join([f"{d:0.3f}" for d in rayDistances[0:10]]))
            print("objectIds", object_ids[0:10])
            for i, object_id in enumerate(object_ids):
                optobj = self.scene.getObjectById(int(object_id))
                if isinstance(optobj, OpticalSurfaceObject):
                    if xp.inner(ArrayUtil.convert(optobj.direction, xp), start_direction) < 0:
                        ior0, ior1 = optobj.ior0, optobj.ior1
                    else:
                        ior0, ior1 = optobj.ior1, optobj.ior0
                    locations.extend([rayDistances[i]] * 2)
                    refractiveIndices.extend([xp.full(shape=(1,), fill_value=ior0), xp.full(shape=(1,), fill_value=ior1)])
                    resultObjects.extend([object_id, object_id])
                elif isinstance(optobj, OpticalVolumeObject):
                    start = rayDistances[i]
                    stop = rayDistances[i + 1] if i + 1 < len(rayDistances) else rangeEnd
                    logger.log(logging.INFO, f"Sampling volume {optobj} from {start} to {stop}")
                    mults = np.linspace(start, stop, 10)
                    volume: OpticalVolumeObject = optobj
                    locations.extend(mults)
                    resultObjects.extend([object_id] * len(mults))
                    refractiveIndices.extend(volume.getRefractiveIndex((start_position[np.newaxis, :] + start_direction[np.newaxis, :] * mults[:,np.newaxis])))
                elif isinstance(optobj, MeshObject):
                    raise NotImplementedError("MeshObject refractive index sampling not implemented yet")
                else:
                    raise NotImplementedError(f"Unknown object type {type(optobj)}")
            self.lineEvaluationDistances = xp.asarray(locations)
            self.lineRefractiveIndices = xp.concatenate(refractiveIndices)
            self.lineObjects = xp.asarray(resultObjects)
        # ---- FOCAL_DISTANCE mode ----
        elif mode == VisualizationMode.FOCAL_DISTANCE:
            num_rays = 3000
            trajectory = []
            lower_bound = xp.full(shape=(num_rays,), fill_value=rangeBegin, dtype=start_position.dtype)
            upper_bound = xp.full(shape=(num_rays,), fill_value=rangeEnd, dtype=start_position.dtype)
            start_objects = self.scene.getActiveSurfaces() + self.scene.getActiveVolumes() + self.scene.getActiveMeshes()
            sp = xp.random.randn(num_rays, 3) * spread
            sp -= (sp @ start_direction.T)[:,np.newaxis] * start_direction[np.newaxis, :]
            all_start_positions = xp.repeat(start_position[xp.newaxis], num_rays, axis=0) + sp
            all_start_directions = xp.repeat(start_direction[xp.newaxis], num_rays, axis=0)
            self.scene.calculateRays(
                all_start_positions,
                all_start_directions,
                start_objects,
                None,
                trajectory,
                lower_bound,
                upper_bound,
                maxVolumeIterations = 3,
                maxBounces = 100,
                xp=xp)

            trajectory = xp.stack(trajectory, axis=1) # shape (num_rays, num_points, 3)

            pyvistadebug = False
            if pyvistadebug:
                #open pyvista-scene that shows all rays in 3d. So for each i (:,i,:) should be a line with multiple points. And for each i, j (i,j,:) should be a point.
                import pyvista as pv
                pl = pv.Plotter()
                trajectory_np = ArrayUtil.convert(trajectory, np)
                for i in range(min(trajectory_np.shape[0],100)):
                    pl.add_lines(np.concatenate((trajectory_np[i,:-1,:], trajectory_np[i, 1:,:]), axis=0), color = np.random.uniform(0,1,3) , width=1)
                #pl.add_mesh(pv.Sphere(radius=0.01, center=ArrayUtil.convert(trajectory_np, np).flatten()), color="red")
                pl.show()


            directions = xp.diff(trajectory, axis=1)
            trajectory_distances = (trajectory - start_position[xp.newaxis, xp.newaxis, :]) @ start_direction # shape (num_rays, num_points)
            indices = xp.ones(num_rays, dtype=xp.uint32)
            all_ray_indices = xp.arange(num_rays)
            intersection_distances = []
            intersection_rear_principal_points = []
            bundleEvaluationDistances = np.linspace(rangeBegin, rangeEnd, num_eval)
            for iDist, d in enumerate(bundleEvaluationDistances):
                to_increase = (trajectory_distances[all_ray_indices, indices] < d) & (indices < trajectory_distances.shape[1] - 1)
                while xp.any(to_increase):
                    indices[to_increase] += 1
                    to_increase = (trajectory_distances[all_ray_indices, indices] < d) & (indices < trajectory_distances.shape[1] - 1)

                current_positions = trajectory[all_ray_indices, indices - 1]
                current_directions = directions[all_ray_indices, indices - 1]
                npc = NearestPointCalculator(position=current_positions,
                                             direction=current_directions,
                                             tolerance=1e-4)

                intersection_distances.append((npc.intersection - start_position) @ start_direction[:, np.newaxis])
                #Calculate back principal points. Use all_start_positions, current_positions and current_directions for this
                npc = NearestPointCalculator(position=current_positions - all_start_positions,
                                       direction=current_directions,)
                intersection_rear_principal_points.append(npc.intersection @ start_direction[:, np.newaxis])


            self.bundleEvaluationDistances = xp.asarray(bundleEvaluationDistances)
            self.bundleIntersectionDistances = xp.concatenate(xp.asarray(intersection_distances))
            self.bundleRearPrincipalPoints = xp.concatenate(xp.asarray(intersection_rear_principal_points))

        # ---- DIVERGENCE mode ----
        elif mode == VisualizationMode.DIVERGENCE:
            num_rays = 1000
            num_eval = 200
            lights = []
            self.scene.getActiveLights(lights)

            bundleEvaluationDistances = np.linspace(rangeBegin, rangeEnd, num_eval)
            for light in lights:
                # Fake ray endpoints and directions
                endpoints = np.random.randn(num_rays, 3)
                enddirs = np.random.randn(num_rays, 3)
                enddirs /= np.linalg.norm(enddirs, axis=1, keepdims=True)

                # variance along direction (simplified stub)
                proj = endpoints @ start_direction
                # group by nearest distance sample
                bins = np.searchsorted(bundleEvaluationDistances, proj)
                variance = np.zeros(num_eval)
                counts = np.zeros(num_eval)
                for i, b in enumerate(bins):
                    if 0 <= b < num_eval:
                        variance[b] += np.linalg.norm(enddirs[i])**2
                        counts[b] += 1
                counts[counts == 0] = 1
                variance = np.sqrt(variance / counts)

                maxv = np.nanmax(variance)